File: code/374.guess-number-higher-or-lower.php
<?php
/*
 * @lc app=leetcode id=374 lang=php
 *
 * [374] Guess Number Higher or Lower
 */

// @lc code=start
/** 
 * The API guess is defined in the parent class.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * public function guess($num){}
 */

class Solution extends GuessGame {
    /**
     * @param  Integer  $n
     * @return Integer
     */
    function guessNumber($n) {
        // 半分ずつ最小と最大を絞っていく
        $min = 1;
        $max = $n;
        $answer = round($max / 2);
        while (true) {
            if ($this->guess($answer) === -1) {
                $max = $answer;
                $answer = round($max / 2);
            } elseif ($this->guess($answer) === 1) {
                $min = $answer;
                $answer = round(($max + $min) / 2);
            } else {
                return $answer;
            }
        }
    }
}
// @lc code=end

File: code/374.guess-number-higher-or-lower.go
package code

/*
 * @lc app=leetcode id=374 lang=golang
 *
 * [374] Guess Number Higher or Lower
 */

// @lc code=start
/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * func guess(num int) int;
 */

func guessNumber(n int) int {
	min := 1
	max := n

	for true {
		ans := (max + min) / 2

		switch guess(ans) {
		case -1:
			max = ans - 1
		case 1:
			min = ans + 1
		default:
			return ans
		}
	}
	return -1
}

// @lc code=end
